### PR TITLE
IF: Sign and verify signature or proposed blocks

### LINKED ISSUE #2142: IF: sign blocks and verify signature of blocks
Currently `hotstuff_integration` branch does not sign or verify signature of proposed blocks after instant finality enabled.

### PR DESCRIPTION
**Pull request by heifner**
Sign `block_id` of proposed blocks and verify signatures on validation nodes.

- Also adds validation of activated protocol features.
- Includes a test that protocol feature activation works after instant_finality is enabled.
- Existing tests cover block signing validation. Many tests failed when I signed with the wrong producer key.

Resolves #2142

### CODE REVIEW

**Review by greg7mdp — APPROVED** [review]
Looks great, nice test for protocol feature activation under savanna.

File: libraries/chain/block_state.cpp [hunk 1]
```diff
@@ -193,4 +203,58 @@ std::optional<quorum_certificate> block_state::get_best_qc() const {
...
+      EOS_ASSERT(res.second, wrong_signing_key, "block signed by same key twice", ("key", *res.first));
+   }
+
+   bool is_satisfied = false;
+   size_t relevant_sig_count = 0;
+
+   std::tie(is_satisfied, relevant_sig_count) = producer_authority::keys_satisfy_and_relevant(keys, valid_block_signing_authority);
+
+   EOS_ASSERT(relevant_sig_count == keys.size(), wrong_signing_key,
+              "block signed by unexpected key",
```
Comment (linh2931) [inline libraries/chain/block_state.cpp:252]: Missing `${signing_keys)` and `${authority}`. You can also log `relevant_sig_count`

File: libraries/chain/block_state.cpp [hunk 2]
```diff
@@ -193,4 +203,58 @@ std::optional<quorum_certificate> block_state::get_best_qc() const {
...
+              ("signing_keys", keys)("authority", valid_block_signing_authority));
+
+   EOS_ASSERT(is_satisfied, wrong_signing_key,
+              "block signatures do not satisfy the block signing authority",
```
Comment (linh2931) [inline libraries/chain/block_state.cpp:256]: The same missing `${signing_keys)` and `${authority}`

File: libraries/chain/block_state.cpp [hunk 3]
```diff
@@ -193,4 +203,58 @@ std::optional<quorum_certificate> block_state::get_best_qc() const {
...
+              ("num_keys", num_keys_in_authority)
+              ("authority", valid_block_signing_authority)
+   );
+
+   std::set<public_key_type> keys;
+   keys.emplace(fc::crypto::public_key( block->producer_signature, block_id, true ));
+
+   for (const auto& s: additional_signatures) {
+      auto res = keys.emplace(s, block_id, true);
+      EOS_ASSERT(res.second, wrong_signing_key, "block signed by same key twice", ("key", *res.first));
```
Comment (linh2931) [inline libraries/chain/block_state.cpp:243]: Missing ${key}

File: libraries/chain/block_state.cpp [hunk 4]
```diff
@@ -193,4 +203,58 @@ std::optional<quorum_certificate> block_state::get_best_qc() const {
...
+
+      emplace_extension(b.block_extensions, additional_block_signatures_extension::extension_id(), fc::raw::pack( additional_signatures ));
+   }
+}
+
+void block_state::sign(const signer_callback_type& signer, const block_signing_authority& valid_block_signing_authority ) {
+   auto sigs = signer( block_id );
+
+   EOS_ASSERT(!sigs.empty(), no_block_signatures, "Signer returned no signatures");
+   block->producer_signature = sigs.back();
```
Comment (linh2931) [inline libraries/chain/block_state.cpp:222]: Maybe add a comment here saying the last is producer_signature and the rest is additional_signatures.

**Review by linh2931 — CHANGES_REQUESTED** [review]
Maybe later add a test for rejecting unsigned blocks?

**Review by linh2931 — APPROVED** [review]

**Review by greg7mdp — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: INTERNALS
summary: Sign block_id of proposed blocks and verify signatures on validation nodes.
Note:end

**Pull request merged**